File: mobile_app/lib/data/models/job_completion_context.dart
class JobCompletionSibling {
  JobCompletionSibling({
    required this.diaryEventId,
    this.officerFullName,
    this.visitStatus,
    this.nextJobState,
    required this.jobReportSubmitted,
    required this.isCurrentVisit,
    required this.visitIsOpen,
  });

  factory JobCompletionSibling.fromJson(Map<String, dynamic> json) {
    return JobCompletionSibling(
      diaryEventId: (json['diary_event_id'] as num).toInt(),
      officerFullName: json['officer_full_name'] as String?,
      visitStatus: json['visit_status'] as String?,
      nextJobState: json['next_job_state'] as String?,
      jobReportSubmitted: json['job_report_submitted'] == true,
      isCurrentVisit: json['is_current_visit'] == true,
      visitIsOpen: json['visit_is_open'] == true,
    );
  }

  final int diaryEventId;
  final String? officerFullName;
  final String? visitStatus;
  final String? nextJobState;
  final bool jobReportSubmitted;
  final bool isCurrentVisit;
  final bool visitIsOpen;
}

class FinishedSiblingChoice {
  FinishedSiblingChoice({
    this.officerFullName,
    this.nextJobState,
  });

  factory FinishedSiblingChoice.fromJson(Map<String, dynamic> json) {
    return FinishedSiblingChoice(
      officerFullName: json['officer_full_name'] as String?,
      nextJobState: json['next_job_state'] as String?,
    );
  }

  final String? officerFullName;
  final String? nextJobState;
}

class JobCompletionContext {
  JobCompletionContext({
    required this.hasMultipleEngineers,
    required this.openVisitCount,
    required this.siblings,
    required this.hasStageConflict,
    required this.distinctChosenStates,
    this.currentJobState,
    required this.isLastEngineerToComplete,
    required this.finishedSiblingChoices,
  });

  factory JobCompletionContext.fromJson(Map<String, dynamic>? json) {
    if (json == null) {
      return JobCompletionContext.empty();
    }
    final raw = json['siblings'];
    final siblings = <JobCompletionSibling>[];
    if (raw is List) {
      for (final e in raw) {
        if (e is Map) {
          siblings.add(
            JobCompletionSibling.fromJson(Map<String, dynamic>.from(e)),
          );
        }
      }
    }
    final statesRaw = json['distinct_chosen_states'];
    final states = <String>[];
    if (statesRaw is List) {
      for (final s in statesRaw) {
        if (s != null && s.toString().trim().isNotEmpty) {
          states.add(s.toString().trim());
        }
      }
    }
    final finishedRaw = json['finished_sibling_choices'];
    final finished = <FinishedSiblingChoice>[];
    if (finishedRaw is List) {
      for (final e in finishedRaw) {
        if (e is Map) {
          finished.add(
            FinishedSiblingChoice.fromJson(Map<String, dynamic>.from(e)),
          );
        }
      }
    }
    return JobCompletionContext(
      hasMultipleEngineers: json['has_multiple_engineers'] == true,
      openVisitCount: (json['open_visit_count'] as num?)?.toInt() ?? 0,
      siblings: siblings,
      hasStageConflict: json['has_stage_conflict'] == true,
      distinctChosenStates: states,
      currentJobState: json['current_job_state'] as String?,
      isLastEngineerToComplete: json['is_last_engineer_to_complete'] == true,
      finishedSiblingChoices: finished,
    );
  }

  factory JobCompletionContext.empty() {
    return JobCompletionContext(
      hasMultipleEngineers: false,
      openVisitCount: 0,
      siblings: const [],
      hasStageConflict: false,
      distinctChosenStates: const [],
      isLastEngineerToComplete: false,
      finishedSiblingChoices: const [],
    );
  }

  final bool hasMultipleEngineers;
  final int openVisitCount;
  final List<JobCompletionSibling> siblings;
  final bool hasStageConflict;
  final List<String> distinctChosenStates;
  final String? currentJobState;
  final bool isLastEngineerToComplete;
  final List<FinishedSiblingChoice> finishedSiblingChoices;

  List<JobCompletionSibling> get otherOpenSiblings => siblings
      .where((s) => !s.isCurrentVisit && s.visitIsOpen)
      .toList();

  List<JobCompletionSibling> get otherSiblingsWithStageChoice => siblings
      .where(
        (s) =>
            !s.isCurrentVisit &&
            s.nextJobState != null &&
            s.nextJobState!.trim().isNotEmpty,
      )
      .toList();
}

File: backend/src/jobCompletionContext.ts
import type { Pool, PoolClient } from 'pg';

type Db = Pool | PoolClient;

export type JobCompletionSibling = {
  diary_event_id: number;
  officer_full_name: string | null;
  visit_status: string | null;
  next_job_state: string | null;
  job_report_submitted: boolean;
  is_current_visit: boolean;
  visit_is_open: boolean;
};

export type FinishedSiblingChoice = {
  officer_full_name: string | null;
  next_job_state: string | null;
};

export type JobCompletionContext = {
  has_multiple_engineers: boolean;
  open_visit_count: number;
  siblings: JobCompletionSibling[];
  has_stage_conflict: boolean;
  distinct_chosen_states: string[];
  /** `jobs.state` before this visit closes the job. */
  current_job_state: string | null;
  /** True when only this engineer's visit is still open on a multi-engineer job. */
  is_last_engineer_to_complete: boolean;
  /** Engineers who already completed their visit and what they chose. */
  finished_sibling_choices: FinishedSiblingChoice[];
};

function visitIsOpenSql(): string {
  return `COALESCE(LOWER(TRIM(REPLACE(COALESCE(d.status, ''), ' ', '_'))), '') NOT IN ('completed', 'cancelled', 'aborted')`;
}

/**
 * Other engineers on the same job may choose different post-visit job stages.
 * Job state applies only after every visit is closed; the last completion wins.
 */
export async function buildJobCompletionContext(
  db: Db,
  jobId: number,
  currentDiaryEventId: number,
): Promise<JobCompletionContext> {
  const res = await db.query<{
    diary_event_id: number;
    officer_full_name: string | null;
    visit_status: string | null;
    next_job_state: string | null;
    job_report_submitted: boolean;
    visit_is_open: boolean;
  }>(
    `SELECT d.id AS diary_event_id,
            o.full_name AS officer_full_name,
            d.status AS visit_status,
            d.next_job_state,
            EXISTS (
              SELECT 1 FROM diary_event_status_logs l
              WHERE l.diary_event_id = d.id AND l.status = 'job_report_submitted'
            ) AS job_report_submitted,
            (${visitIsOpenSql()}) AS visit_is_open
     FROM diary_events d
     LEFT JOIN officers o ON o.id = d.officer_id
     WHERE d.job_id = $1
       AND COALESCE(LOWER(TRIM(REPLACE(COALESCE(d.status, ''), ' ', '_'))), '') <> 'aborted'
     ORDER BY d.start_time ASC NULLS LAST, d.id ASC`,
    [jobId],
  );

  const siblings: JobCompletionSibling[] = res.rows.map((r) => ({
    diary_event_id: Number(r.diary_event_id),
    officer_full_name: r.officer_full_name,
    visit_status: r.visit_status,
    next_job_state: r.next_job_state,
    job_report_submitted: !!r.job_report_submitted,
    is_current_visit: Number(r.diary_event_id) === currentDiaryEventId,
    visit_is_open: !!r.visit_is_open,
  }));

  const openVisitCount = siblings.filter((s) => s.visit_is_open).length;
  const hasMultipleEngineers = siblings.length > 1;

  const chosenStates = new Set<string>();
  for (const s of siblings) {
    if (s.next_job_state && String(s.next_job_state).trim()) {
      chosenStates.add(String(s.next_job_state).trim());
    }
  }

  const jobRes = await db.query<{ state: string | null }>(
    `SELECT state FROM jobs WHERE id = $1`,
    [jobId],
  );
  const currentJobState =
    jobRes.rows[0]?.state != null && String(jobRes.rows[0].state).trim()
      ? String(jobRes.rows[0].state).trim()
      : null;

  const currentSibling = siblings.find((s) => s.is_current_visit);
  const isLastEngineerToComplete =
    hasMultipleEngineers &&
    openVisitCount === 1 &&
    currentSibling?.visit_is_open === true;

  const finishedSiblingChoices: FinishedSiblingChoice[] = siblings
    .filter((s) => !s.is_current_visit && !s.visit_is_open)
    .map((s) => ({
      officer_full_name: s.officer_full_name,
      next_job_state: s.next_job_state,
    }));

  return {
    has_multiple_engineers: hasMultipleEngineers,
    open_visit_count: openVisitCount,
    siblings,
    has_stage_conflict: chosenStates.size > 1,
    distinct_chosen_states: [...chosenStates],
    current_job_state: currentJobState,
    is_last_engineer_to_complete: isLastEngineerToComplete,
    finished_sibling_choices: finishedSiblingChoices,
  };
}
